### ISSUE TITLE
Re-add `links` in model definition with v0.7.x format?

### ISSUE THREAD

**Issue by jgu2** [calliope-project/calliope#668]
### Description
In model definition of Calliope `v0.7.x`, the `links` key got removed and defined as transmission technologies under `techs` in yaml, the doc - https://calliope.readthedocs.io/en/latest/migrating/#links-transmission-links-defined-in-techs

This change can let users get `links` items from model definition easier. And, it looks more natural with `links` siting along `techs`.

### Proposed Change

Add `links` key and allow users to define transmission technologies, for example,

Current Yaml
```bash
techs:
  ccgt:
    name: "Combined cycle gas turbine"
    ...

  csp:
    name: "Concentrating solar power"
    ...

  battery:
    name: "Battery storage"
    ...

  demand_power:
    name: "Power demand"
    ...

  region1_to_region2:
    from: region1
    to: region2
    name: "AC power transmission"
    ...
  region1_to_region1_1:
    from: region1
    to: region1_1
    inherit: free_transmission
  region1_to_region1_2:
    from: region1
    to: region1_2
    inherit: free_transmission
  region1_to_region1_3:
    from: region1
    to: region1_3
    inherit: free_transmission
```

After the change,
```bash

techs:
  ccgt:
    name: "Combined cycle gas turbine"
    ...

  csp:
    name: "Concentrating solar power"
    ...

  battery:
    name: "Battery storage"
    ...

  demand_power:
    name: "Power demand"
    ...

links:
  region1_to_region2:
    from: region1
    to: region2
    name: "AC power transmission"
    ...
  region1_to_region1_1:
    from: region1
    to: region1_1
    inherit: free_transmission
  region1_to_region1_2:
    from: region1
    to: region1_2
    inherit: free_transmission
  region1_to_region1_3:
    from: region1
    to: region1_3
    inherit: free_transmission
```

### Minimum Solution

To make this change to be compatible with current schema validation,
In `src.calliope.model.py`, could add following in to `def _init_from_model_def_dict` before `validate_dict`,

```python
if "links" in model_definition:
    model_definition["techs"].update(model_definition.pop("links"))
```

That is,

```python

    def _init_from_model_def_dict(
        self,
        model_definition: calliope.AttrDict,
        applied_overrides: str,
        scenario: str | None,
        data_source_dfs: dict[str, pd.DataFrame] | None = None,
    ) -> None:
        """Initialise the model using pre-processed YAML files and optional dataframes/dicts.

        Args:
            model_definition (calliope.AttrDict): preprocessed model configuration.
            applied_overrides (str): overrides specified by users
            scenario (str | None): scenario specified by users
            data_source_dfs (dict[str, pd.DataFrame] | None, optional): files with additional model information. Defaults to None.
        """
        
        
        # Pop the `links` and add items into `tech`, making it compatible with current schema.
        if "links" in model_definition:
            model_definition["techs"].update(model_definition.pop("links"))


        # First pass to check top-level keys are all good
        validate_dict(model_definition, CONFIG_SCHEMA, "Model definition")

```

### Version

[v0.7.0.dev3](https://github.com/calliope-project/calliope/releases/tag/v0.7.0.dev3)

**Comment by brynpickering** [comment]
The reason for the change was because links were being turned into `techs` in the pre-processed Calliope model in v0.6 _anyway_. That is, if you defined a link "ac_transmission" from "A" to "B" there would be an entry "A::ac_transmission:B" in the model data dimension `techs`. 

The change makes it clearer to the user that they are effectively defining technologies when they define transmission links and that they can then expect to find those technologies on the dimension `techs` in their Calliope model inputs and results.

We tried to work out a way to have them on their own dimension (`links`), but this causes problems with defining parameters and decision variables that are shared with non-transmission technologies (e.g. `flow_out_eff`, `flow_cap`). See [this comment](https://github.com/calliope-project/calliope/issues/324#issuecomment-1756987730) for more detail. If you have an idea of how we might achieve that, I'm very open to trying to implement it!

**Comment by jgu2** [comment]
Thanks for the note. 

Yeah, that's why we are trying to make `links` to be optional, if `links` appears in yaml, then merge it into `techs` to align with what current model schema look like. This change is trying to facilitate our current gradient modeling at NREL.

@jmorrisnrel, do you have any input regarding this?

**Comment by jmorrisnrel** [comment]
Yeah this request would primarily be to help facilitate results processing both for visualizations/analysis and for automating multi-year modeling buildouts. We can easily shift the old behavior of transmission "techs" from `techs` to `tech_groups`, but having to place the individual transmission lines in `techs` means we have to add extra processing to distinguish actual tech definitions in `techs` from individual transmission lines when processing especially when looking for lines that may have been built in a previous year. Having an optional `links` tag to specify those in makes the processing cleaner and the YAML a little more readable/familiar to users of v0.6.

The change we implemented doesn't actually impact any of the problem formulation under the hood, instead just looking for `links` and doing a dictionary update of the model dict after it is loaded to add the content of `links` to `techs`. Using it would abstract the YAML definition away from cleanly matching the actual dimensions of the underlying model which may not align with your goals so happy to discuss if there is another alternative.

**Comment by jmorrisnrel** [comment]
Because we have discrete containerized compute jobs for each solve (with several chained together for a multiyear sequential buildout), we can't easily keep model results in memory between model runs. We are writing all results to disk (currently in the default CSV export format) and create an "output" YAML with all of the input configurations (model, techs, locations, ect.) with inline `results` sections at the `loc_tech`/`link` level in v0.6. This also lets users who aren't as comfortable with python/xarray to review the results in detail in a more human readable format. We could move the transmission `techs` block to another file and rewrite this code to split the files back out, but having the transmission "links" in a separate section makes this a little more readable for our users. 
![image](https://github.com/user-attachments/assets/0d12be6e-943e-4395-a805-199b71225394)

Our current process for sequential year runs primarily works with this output YAML file as well as the input YAMLs for future years in dictionary form. We load in the results and using those update the future year inputs. Having the full set of inputs right there in the same dict lets us easily copy over any characteristics from the previous run (performance, O&M, ect.) that we need to correctly represent the technology built in that year in the future year run. We'll need to split the outputs YAML into separate files or add some code that either reads one of the input files or hunts through the inheritance in the YAML to determine what `techs` are transmission so we can determine what was built to copy over to the new year. With a `links` section we can just loop through that instead which simplifies things. We can certainly adjust our code and likely will do a rewrite of how we process results in the future, but it looked like a relatively small change to add optional support for `links` back in v0.7 with a quick dict update.

**Comment by irm-codebase** [comment]
Reviving this thread @brynpickering 
After using the code a bit, I think I see the point of re-adding links (or some type of `edges` dimension) support into the formulation.

Since `templates:` are not supported within tables (due to limitations in YAML construction provenance), you need to define a lot of extra technologies, leading to the technology array size growing quite quickly. It also produces very long and unintuitive YAML files (see the attached example).
[hv_links.txt](https://github.com/user-attachments/files/32243062/hv_links.txt)

I suspect this could also lead to small efficiency improvements here and there by making our `model.inputs` less sparse.

**Comment by brynpickering** [comment]
We return to the original problem, though: defining a link/edge either
needs a new model dimension or it needs creating techs that represent each
end of an edge. Creating new techs from some kind of string concatenation
on user definition in YAML/CSV was always a pain to deal with and there's
no real way of getting away from it without some kind of overhaul on the
data storage side.

_if_ we had the separate concept of links, I'd say they would be
complementary to defining transmission techs directly and would just be a
helper to halve the number of definitions made for transmission. There
would need to be some explicit concatenation rule that the user defines so
they understand what their link will become (two separate techs) in the
input data.

**Comment by irm-codebase** [comment]
Yeah, it is quite a pickle... There must be a way to deal with this at the root of the problem, though. I'd prefer if we avoid going back to v0.6's method of silently 'splitting' links into two technologies. Our math and backend code is mature enough to support something better. 

At the end of the day, Calliope follows graph theory to a certain degree. There must be a way to adapt our mathematics and data handling to accommodate the concept of constrained "edges", and to do so in a way that requires less nuance in how we extract and store this data. I suspect this means handling edges with a different 'capacity' variable (e.g., `link_cap` or `edge_cap`), to finally decouple transmission from the need to split it into two technologies. Then:
- node secluded technologies only use `flow_cap` and `storage_cap`, which can transform or store `carriers` (which essentially always represented a per-node sub-network of unconstrained edges).
- `link_cap` (or whichever name we pick) handles constrained edge connections between these sub-networks.

This is merely based on gut feelings, though. We'll have to find the time to think of a good design, to implement this well and solve the root of the problem.

**Comment by brynpickering** [comment]
The problem we faced every time we discussed this previously is that the
number of parameters and variables shared between links and other techs is
quite high. `flow_cap` is one, but so is `flow_out`, `flow_in`, `active`,
`flow_out_eff`, `flow_in_eff`, and perhaps more.

Handling it in the math would be feasible, with separate sets of variables
and constraints just for links. The inputs are the bigger problem as we
can't currently have two of the same names parameters indexed over
different dimensions in the same dataset. So it would possibly require
splitting the dataset into multiple ones.

I also don't actually quite know where the overall benefit is on the
definition side. You still have to define each edge of the graph. The
current approach is also just one defined tech per edge. Is it the issue of
having to give them distinct names?

**Comment by irm-codebase** [comment]
We should definitely avoid introducing ambiguities. Having two parameters with different dimensions would spell trouble in other areas. Whatever we do should lead to simpler and more resilient code for both us and the user. Introducing a bunch of new parameters / variables would be cumbersome, but that might be fine if it's just 4 or 5 and it leads to significant reductions in code complexity. Looking at other models, seems like PyPSA has [lines and links ](https://docs.pypsa.org/latest/api/components/types/links/#pypsa.components.Links) as edge-like options.

On the definition side, I'd say that having to create a bunch of new unique names is quite annoying, yeah. On the user side that'd be the biggest gain. I do not want the solution to re-introduce the parsing headaches you mentioned above, which is why I proposed to define edges as their own thing.

**Comment by brynpickering** [comment]
But PyPSA stores them as individual named entities in its data model (one
row per named link, one row per numbered lines with the numbers possibly
automatically generated). So I don't think it's actually so different to
what we do, certainly once it's in the internal data model.

I still don't understand how defining edges is different to what we
currently do. Are you saying that we might have a single technology (e.g.
AC line) and a list of all its edges? The difficulty would then be if you
want different properties per edge where those properties overlap with
other technology classes (e.g. investment cost) as you would have to define
new parameters for every one of them, indexed over (tech, edge) rather than
(tech, node). And then what do you call the `edge`? Ideally it is a tuple
of (node, node) but that is a big no-no in xarray (and pandas) so then you
need to name edges and we're therefore back at stage 1. You have as many
names you need to define for tech*edges as you do currently for tech*nodes.

**Comment by brynpickering** [comment]
I should add: I want to find a solution to this. I also don't like it as it
currently is. But it's something we've thought about a lot in the past and
not yet found anything satisfactory.
